api/admin/announcement_list_validator.py

```diff
@@ -0,0 +1,155 @@
+                _("Too many announcements: maximum is %(maximum)d",
+                  maximum=self.maximum_announcements)
+            )
+            
+        seen_ids = set()
+        for announcement in announcements:
+            validated = self.validate_announcement(announcement)
+            if isinstance(validated, ProblemDetail):
+                return validated
+            id = validated['id']
+            if id in seen_ids:
+                return INVALID_INPUT.detailed(_("Duplicate announcement ID: %s" % id))
+            seen_ids.add(id)
+            validated_announcements.append(validated)
+        return validated_announcements
+
+    def validate_announcement(self, announcement):
+        validated = dict()
+        if not isinstance(announcement, dict):
+            return INVALID_INPUT.detailed(
+                _("Invalid announcement format: %(announcement)r", announcement=announcement)
+            )
+
+        validated['id'] = announcement.get('id', unicode(uuid.uuid4()))
+
+        for required_field in ('start', 'content'):
+            if not required_field in announcement:
+                return INVALID_INPUT.detailed(
+                    _("Missing required field: %(field)s", field=required_field)
+                )
+
+        # Validate the content of the announcement.
+        content = announcement['content']
+        content = self.validate_length(
+            content, self.minimum_announcement_length, self.maximum_announcement_length
+        )
+        if isinstance(content, ProblemDetail):
+            return content
+        validated['content'] = content
+
+        # Validate the dates associated with the announcement
+        today = datetime.date.today()
+
+        start = self.validate_date(
+            'start', announcement.get('start', today), minimum=today
+        )
+        if isinstance(start, ProblemDetail):
+            return start
+        validated['start'] = start
+
+        default_finish = start + datetime.timedelta(days=self.default_duration_days)
+        day_after_start = start + datetime.timedelta(days=1)
+        finish = self.validate_date(
+            'finish',
+            announcement.get('finish', default_finish),
+            minimum=day_after_start
+        )
+        if isinstance(finish, ProblemDetail):
+            return finish
+        validated['finish'] = finish
+
+        # That's it!
+        return validated
+
+    @classmethod
+    def validate_length(self, value, minimum, maximum):
+        """Validate the length of a string value.
+
+        :param value: Proposed value for a field.
+        :param minimum: Minimum length.
+        :param maximum: Maximum length.
+
+        :return: A ProblemDetail if the validation fails; otherwise `value`.
+        """
+        if len(value) < minimum:
+            return INVALID_INPUT.detailed(
+                _('Value too short (minimum length is %(size)d characters): %(value)s',
```

Comment:
I think it would be more helpful for the error message to give the length of the proposed value, rather than the actual value.